File: core/esmf-aspect-meta-model-java/src/main/java/org/eclipse/esmf/metamodel/impl/DefaultAspect.java
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for additional
 * information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */
package org.eclipse.esmf.metamodel.impl;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.StringJoiner;

import org.eclipse.esmf.metamodel.Aspect;
import org.eclipse.esmf.metamodel.Event;
import org.eclipse.esmf.metamodel.Operation;
import org.eclipse.esmf.metamodel.Property;
import org.eclipse.esmf.aspectmodel.loader.MetaModelBaseAttributes;
import org.eclipse.esmf.aspectmodel.visitor.AspectVisitor;

public class DefaultAspect extends ModelElementImpl implements Aspect {
   private final List<Property> properties;
   private final List<Operation> operations;
   private final List<Event> events;
   private final boolean isCollectionAspect;

   public DefaultAspect( final MetaModelBaseAttributes metaModelBaseAttributes,
         final List<Property> properties,
         final List<Operation> operations,
         final List<Event> events,
         final boolean isCollectionAspect ) {
      super( metaModelBaseAttributes );
      this.properties = new ArrayList<>( properties );
      this.operations = new ArrayList<>( operations );
      this.events = new ArrayList<>( events );
      this.isCollectionAspect = isCollectionAspect;
   }

   /**
    * A list of Properties exposed by the Aspect.
    *
    * @return the properties.
    */
   @Override
   public List<Property> getProperties() {
      return properties;
   }

   /**
    * A list of Operations exposed by the Aspect.
    *
    * @return the operations.
    */
   @Override
   public List<Operation> getOperations() {
      return operations;
   }

   /**
    * A list of Events provided by the Aspect.
    *
    * @return the events.
    */
   @Override
   public List<Event> getEvents() {
      return events;
   }

   /**
    * Accepts an Aspect visitor
    *
    * @param visitor The visitor to accept
    * @param <T> The result type of the traversal operation
    * @param <C> The context of the visitor traversal
    */
   @Override
   public <T, C> T accept( final AspectVisitor<T, C> visitor, final C context ) {
      return visitor.visitAspect( this, context );
   }

   @Override
   public boolean isCollectionAspect() {
      return isCollectionAspect;
   }

   @Override
   public String toString() {
      return new StringJoiner( ", ", DefaultAspect.class.getSimpleName() + "[", "]" )
            .add( "properties=" + properties )
            .add( "operations=" + operations )
            .add( "events=" + events )
            .add( "isCollectionAspect=" + isCollectionAspect )
            .toString();
   }

   @Override
   public boolean equals( final Object o ) {
      if ( this == o ) {
         return true;
      }
      if ( o == null || getClass() != o.getClass() ) {
         return false;
      }
      if ( !super.equals( o ) ) {
         return false;
      }
      final DefaultAspect that = (DefaultAspect) o;
      return isCollectionAspect == that.isCollectionAspect
            && Objects.equals( properties, that.properties )
            && Objects.equals( operations, that.operations )
            && Objects.equals( events, that.events );
   }

   @Override
   public int hashCode() {
      return Objects.hash( super.hashCode(), properties, operations, events, isCollectionAspect );
   }
}

File: core/esmf-aspect-model-document-generators/src/test/resources/ts/snapshots/esmf/aspect-meta-model/default-aspect.ts
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for
 * additional information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import {Base, DefaultCollection, Event, Operation, Property} from './index';
import {BaseMetaModelElement} from './base';
import {HasProperties} from './has-properties';
import {ModelVisitor} from '../visitor/model-visitor';

export interface Aspect extends BaseMetaModelElement, HasProperties {
    operations: Array<Operation>;
    events: Array<Event>;
    isCollectionAspect?: boolean;
}

export class DefaultAspect extends Base implements Aspect {
    constructor(
        metaModelVersion: string,
        aspectModelUrn: string,
        name: string,
        private _properties: Array<Property> = [],
        private _operations: Array<Operation> = [],
        private _events: Array<Event> = [],
        private _isCollectionAspect: boolean = false
    ) {
        super(metaModelVersion, aspectModelUrn, name);
        const collectionProperty = _properties.find(property => property.characteristic instanceof DefaultCollection);
        if (collectionProperty && collectionProperty.name === 'items') {
            this._isCollectionAspect = true;
        }
    }

    public getProperty(name: string): Property {
        if (!this._properties) {
            return undefined;
        }
        return this._properties.find(property => property.name === name);
    }

    public get properties(): Array<Property> {
        return this._properties;
    }

    public get operations(): Array<Operation> {
        return this._operations;
    }

    public get isCollectionAspect(): boolean {
        return this._isCollectionAspect;
    }

    get events(): Array<Event> {
        return this._events;
    }

    set events(value: Array<Event>) {
        this._events = value;
    }

    public accept<T, U>(visitor: ModelVisitor<T, U>, context: U): T {
        return visitor.visitAspect(this, context);
    }
}
